feat(order): 주문상세조회, 기간, userId, storeId 로 검색가능한 주문전체조회 기능 추가

yawning5 opened the pull request:
## OrderController 

### .getOrder
- 검색 옵션(dto)과 페이징 객체를 받아오게 되었습니다

### 공통
- 각 요청 호출 및 종료시 log 가 출력되도록 하였습니다
- Service 에 넘겨주는 인자가 
userId -> userDetails 로 변경되었습니다
  (유저 권한도 서비스로 넘겨주기 위함)

## OrderResponseDto
- 응답필드 orderStatus 를 추가하였습니다

## OrderSearchOptionDto
- 컨트롤러에서 받아오는 검색조건 컬럼들을 한번에 받아올 수 있는 Dto 입니다
  (Param 으로 받아오니 Controller 가 너무 두꺼워졌고 검색조건이 많아짐에 따라 dto 로 관리하는 것이 좋다고 생각되었습니다)

## Order
- orderStatus, orderMenu 필드의 초기화를 삭제하였습니다
  (Builder 로만 생성되기 때문에 Builder 내부로 이관)

- address 필드의 nullable false 추가

## OrderRepository
- OrderRepository JPQL 을 사용한 검색쿼리를 작성하였습니다 오더전체조회시에는 쿼리가 총 3번 나가게 됩니다

(

1번 쿼리 orderId 를 검색조건에 맞게 Page 해서 가져옴

2번 쿼리 검색조건에 맞는 total size 를 찾는 쿼리

3번 각 orderId 에 맞는 연관 테이블을 조회하는 쿼리

)

---

## Test

### OrderRepositoryTest
- 단건조회에 관한 테스트코드를 작성하였습니다

### OrderServiceTest
- OrderService 가 받는 인자가 변함에 따라 동기화하였습니다

---

## 이후 예정사항

1. 튜터님 피드백에 따른 Service 계층 리팩토링
2. 권한에 따른 동작 분화 추가
3. 단위 테스트코드 작성 및 이전 테스트코드 리팩토링

예정순서는 팀내 회의에 따라 변경될 수 있습니다 감사합니다

yawning5 committed:
feat(order): 단건조회와 전체조회 추가

기능
 단건조회: 단건에 대한 상세조회

 전체조회: 테이블에 있는 모든 주문목록 조회 페이징 기본 10

 repo테스트와 컴파일 모두 성공한 커밋입니다
---

공통
권한에 따른 동작 분화 추가예정

전체조회
검색기능 추가중입니다

yawning5 committed:
feat(order): 단건조회와 전체조회 추가

기능
 단건조회: 단건에 대한 상세조회

 전체조회: 테이블에 있는 모든 주문목록 조회 페이징 기본 10

---

공통
권한에 따른 동작 분화 추가예정

전체조회
검색기능 추가중입니다

yawning5 committed:
Merge remote-tracking branch 'origin/feat/order/getOrder' into feat/order/getOrder

yawning5 committed:
Merge branch 'dev' into feat/order/getOrder

yawning5 committed:
feat(order): 전체조회시 기간, userId, storeId 에 따른 검색기능 추가

기존 order 관련 테스트 통과확인 하였고
컴파일 에러 없는 부분도 확인하였습니다

---

이후 예정사항

1. 튜터님 피드백에 따른 Service 계층 리팩토링
2. 권한에 따른 동작 분화 추가
3. 단위 테스트코드 작성 및 이전 테스트코드 리팩토링

bbangjae commented on src/main/java/com/sparta/tdd/domain/order/repository/querydsl/OrderRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,121 @@
+
+    @Override
+    public Optional<Order> findDetailById(UUID id) {
+        QOrder o = QOrder.order;
+        QUser u = QUser.user;
+        QStore s = QStore.store;
+        QPayment p = QPayment.payment;
+        QOrderMenu om = QOrderMenu.orderMenu;
+        QMenu m = QMenu.menu;
```
개인적으로 별칭대신 테이블 명을 적어주는 게 가독성 면에서는 좋다고 생각합니다
(개인적인 생각입니다 ㅎㅎ)

yawning5 replied:
감사합니다 반영하도록하겠습니다

bbangjae commented on src/main/java/com/sparta/tdd/domain/order/repository/querydsl/OrderRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,121 @@
+            total = 0L;
+        }
+
+        return new PageImpl<>(ids, pageable, total);
```
`PageableExecutionUtils` 도 있습니다 참고해주세요!(count 쿼리 성능 개선 부분)

https://junior-datalist.tistory.com/342

yawning5 replied:
확인해보겠습니다 정보 감사합니다!!

yawning5 replied:
적용부분은 제가 공부를 좀 하고나서 이후 pr 에서 적용하도록 하겠습니다

bbangjae reviewed (approved):
고생하셨습니다!

dnjsals45 commented on src/main/java/com/sparta/tdd/domain/order/repository/OrderRepository.java:
```diff
@@ -1,11 +1,45 @@
 package com.sparta.tdd.domain.order.repository;
 
 import com.sparta.tdd.domain.order.entity.Order;
+import java.util.Collection;
+import java.util.List;
 import java.util.UUID;
 import org.springframework.data.jpa.repository.JpaRepository;
+import org.springframework.data.jpa.repository.Query;
 import org.springframework.stereotype.Repository;
 
 @Repository
-public interface OrderRepository extends JpaRepository<Order, UUID> {
+public interface OrderRepository extends JpaRepository<Order, UUID>, OrderRepositoryCustom {
 
+//    @Query("""
+//            select distinct o
+//            from Order o
+//            left join fetch o.user u
+//            left join fetch o.store s
+//            left join fetch o.payment p
+//            left join fetch o.orderMenuList om
+//            left join fetch om.menu m
+//            where o.id = :id
+//        """)
+//    Optional<Order> findDetailById(UUID id);
+
+//    @Query("""
+//          select o.id
+//          from Order o
+//          order by o.createdAt desc
+//        """)
+//    Page<UUID> findPageIds(Pageable pageable, Long targetUserId, LocalDateTime start,
+//        LocalDateTime end, UUID targetStoreId);
```
이 코드는 주석된채로 push하신 이유가 있을까요?

yawning5 replied:
제가 쿼리Dsl 이 익숙하지 않기도해서 문제가 생겼을때 주석만 풀면 같은 기능으로 JPQL 대체를 빠르게 할 수 있게 두려고 했습니다
없애는 편이 좋을까요..

dnjsals45 replied:
개인적인 의견으로는 PR에 작성된 코드에서 별도의 코멘트가 없이 불필요한 코드가 존재하면 해당 파일 관련해서 Git 충돌이 발생하거나 팀원들의 궁금증을 유발할 수 있을 것 같다고 생각합니다. 
하지만, 아직 배포가 이루어지거나 최종적으로 코드를 완성하고 유지보수 및 확장 단계가 아니기 때문에 무조건 없어야 한다 까지는 아니더라도 주석으로 설정해두신 이유정도는 메소드 위쪽에 작성해두시면 더 좋지 않을까 생각합니다.

yawning5 replied:
작성해서 새로 커밋 올리도록 하겠습니다!

yawning5 replied:
이유 작성 및 주석처리된 코드는 아래로 위치 조정하였습니다!

dnjsals45 commented on src/main/java/com/sparta/tdd/domain/order/repository/querydsl/OrderRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,121 @@
+        QOrderMenu om = QOrderMenu.orderMenu;
+        QMenu m = QMenu.menu;
+
+        Order result = query
+            .selectFrom(o)
+            .distinct()
+            .leftJoin(o.user, u).fetchJoin()
+            .leftJoin(o.store, s).fetchJoin()
+            .leftJoin(o.payment, p).fetchJoin()
+            .leftJoin(o.orderMenuList, om).fetchJoin()
+            .leftJoin(om.menu, m).fetchJoin()
+            .where(o.id.eq(id))
+            .fetchOne();
```
삭제된 주문의 상세 정보 조회에 대해서는 어떻게 처리하는 것이 좋을까요? 그리고 다른 사람의 주문 건에 대해서는 조회하지 못해야할 것 같다는 생각도 있는데 한 번 정리를 해봐야겠네요(MASTER 권한을 제외하고?)

yawning5 replied:
그 부분은 서비스단에서 처리를 할 생각이긴 합니다

userId = null 로 오면 managert, master 권한 아니면 예외를 던지고 하는 식으로요
customer 같은 경운는 넘어온 userId 와 쿼리의 userId 가 동일해야 로직이 작동하는 식으로

튜터님 피드백에 따른 리팩토링 이후 추가하려고 생각중에 있었습니다

dnjsals45 commented on src/main/java/com/sparta/tdd/domain/order/repository/querydsl/OrderRepositoryCustomImpl.java:
```diff
@@ -0,0 +1,121 @@
+        }
+
+        List<OrderSpecifier<?>> specifiers = new ArrayList<>();
+        for (Sort.Order sort_order : sort) {
```
팀 컨벤션에 따라 sort_order -> sortOrder 로 변경되어야 할  것 같습니다.

p990805 reviewed (approved):
위에 전체적으로 log.Info가 있는데 후에 나중에 지우시는걸까요??

yawning5 committed:
Merge branch 'dev' into feat/order/getOrder

# Conflicts:
#	src/main/java/com/sparta/tdd/domain/order/repository/OrderRepository.java

yawning5 committed:
fix(order): 테이블명 별칭이 아닌 qOrder 식으로 수정
Order 예약어로 겹쳐 q를 앞에 붙여 통일

sort_order 부분 sortOrder 로 컨벤션에 맞게 변경

컨플릭트 부분
코드삭제 없이 OrderRepository 에 추가하는 식으로 병합하였습니다

orderRepository 부분 주석은 의견주시면 (삭제 or 유지) 하도록 하겠습니다

dnjsals45 commented:
> 위에 전체적으로 log.Info가 있는데 후에 나중에 지우시는걸까요??

저도 이 부분 관련해서는, 디버깅 용도라면 컨트롤러 진입 시점과 나가는 시점에서  AOP등을 활용해서 추적하도록 설정해도 좋을 것 같습니다

yawning5 commented:
AOP 관련한 코드는 order 도메인이 어느정도 제 기능을 다하면,, 추가해보도록 하겠습니다..

yawning5 committed:
fix(order): OrderRepository 부분 주석 처리된 코드 아래로 위치이동

주석처리한 코드를 남겨둔 이유를 추가하였습니다

dnjsals45 reviewed (approved):
고생하셨습니다!
